astropy/table/tests/test_column.py

```diff
@@ -583,3 +584,142 @@ def test_string_truncation_warning():
+    assert isinstance(c[:0], table.Column)
+    assert np.all(c[:2] == np.array([uba, 'def']))
+
+    assert isinstance(c[:], table.Column)
+    assert c[:].dtype.char == 'S'
+
+    # Array / list comparisons
+    if not six.PY2:
+        assert np.all(c == [uba, 'def'])
+
+    ok = c == [uba8, b'def']
+    assert type(ok) is type(c.data)
+    assert ok.dtype.char == '?'
+    assert np.all(ok)
+
+    assert np.all(c == np.array([uba, u'def']))
+    if not six.PY2:
+        assert np.all(c == np.array([uba8, b'def']))
+
+    # Scalar compare
+    cmps = (uba8,) if six.PY2 else (uba, uba8)
+    for cmp in cmps:
+        ok = c == cmp
+        assert type(ok) is type(c.data)
+        assert np.all(ok == [True, False])
+
+
+def test_col_unicode_sandwich_unicode():
+    """
+    Sanity check that Unicode Column behaves normally.
+    """
+    uba = 'ba' if six.PY2 else u'bä'
+    uba8 = uba.encode('utf-8')
+
+    c = table.Column([uba, 'def'], dtype='U')
+    assert c[0] == uba
+    assert isinstance(c[:0], table.Column)
+    assert isinstance(c[0], six.text_type)
+    assert np.all(c[:2] == np.array([uba, 'def']))
+
+    assert isinstance(c[:], table.Column)
+    assert c[:].dtype.char == 'U'
+
+    ok = c == [uba, 'def']
+    assert type(ok) == np.ndarray
+    assert ok.dtype.char == '?'
+    assert np.all(ok)
+
+    # In PY2 unicode is equal to bytestrings but not in PY3
+    if six.PY2:
+        assert np.all(c == [uba8, b'def'])
+    else:
+        assert np.all(c != [uba8, b'def'])
+
+
+def test_masked_col_unicode_sandwich():
+    """
+    Create a bytestring MaskedColumn and ensure that it works in Python 3 in
+    a convenient way like in Python 2.
+    """
+    c = table.MaskedColumn([b'abc', b'def'])
+    c[1] = np.ma.masked
+    assert isinstance(c[:0], table.MaskedColumn)
+    assert isinstance(c[0], str)
+
+    assert c[0] == 'abc'
+    assert c[1] is np.ma.masked
+
+    assert isinstance(c[:], table.MaskedColumn)
+    assert c[:].dtype.char == 'S'
+
+    ok = c == ['abc', 'def']
+    assert ok[0] == True
+    assert ok[1] is np.ma.masked
+    #assert c == [b'abc', b'def']
+    #assert c == np.array([u'abc', u'def'])
+    #assert c == np.array([b'abc', b'def'])
+
+    for cmp in (u'abc', b'abc'):
+        ok = c == cmp
+        assert type(ok) is np.ma.MaskedArray
+        assert ok[0] == True
+        assert ok[1] is np.ma.masked
+
+
+@pytest.mark.parametrize('Column', (table.Column, table.MaskedColumn))
+def test_unicode_sandwich_set(Column):
+    """
+    Test setting
+    """
+    uba = 'ba' if six.PY2 else u'bä'
+
+    c = Column([b'abc', b'def'])
+
+    c[0] = b'aa'
+    assert np.all(c == [u'aa', u'def'])
+
+    c[0] = uba  # a-umlaut is a 2-byte character in utf-8, test fails with ascii encoding
+    assert np.all(c == [uba, u'def'])
+    assert c.pformat() == [u'None', u'----', '  ' + uba, u' def']
+
+    c[:] = b'cc'
+    assert np.all(c == [u'cc', u'cc'])
+
+    c[:] = uba
+    assert np.all(c == [uba, uba])
```

Comment:
There is no guarantee of "proper" truncation in the sense that you can easily wind up with an ill-formed utf-8 string.  But we do know that a warning will be emitted in this case.  I don't think this needs a separate test because #5624 has good tests, and we know that the encoded version of `u'äää'` will be longer than 3 bytes.  So this would just be applying a different set of bytes to the #5624 testing.